File: Sources/Heeler/Pairing/SSHPairingConnector.swift
// @preconcurrency: Citadel predates strict concurrency; see SSHTransport.
@preconcurrency import Citadel
import CryptoKit
import Foundation
import NIOCore
@preconcurrency import NIOSSH
import Synchronization

/// The production pairing ceremony client (ADR 0007): a dedicated one-shot
/// SSH client beside `SSHTransport`, not inside it. Two short-lived
/// connections per ceremony — bootstrap (authenticate + enroll) and the
/// verified Device Key reconnect — with the host key pinned from the Pairing
/// Code on both; no TOFU prompt, no known-hosts store.
struct SSHPairingConnector: PairingConnector {
    /// TCP-establishment budget per candidate address, so one unreachable
    /// interface fails over quickly. Citadel caps the subsequent handshake
    /// and authentication at 10 seconds on its own.
    var perAddressTimeout: Duration = .seconds(4)
    /// Deadline for the whole Enrollment exec exchange. Generous: the accept
    /// entrypoint edits authorized_keys under a lock with its own 5s bound.
    var enrollTimeout: Duration = .seconds(20)
    /// Comment on the submitted Device Key line; the plugin popup shows it
    /// as the enrolled device's label. Must be printable ASCII — the accept
    /// entrypoint rejects anything else as `invalid_key`.
    var deviceKeyComment: String = "heeler"

    func pair(
        code: PairingCode,
        deviceKey: DeviceKey,
        onStep: @escaping @Sendable (PairingStep) -> Void
    ) async throws -> PairingResult {
        guard let bootstrap = code.bootstrap else {
            // A config-only Pairing Code is this ceremony minus the
            // bootstrap connection: the Device Key must already be
            // authorized on the Host (pasted manually), so the first
            // connect doubles as the verification.
            onStep(.reach)
            let reached = try await reach(
                code: code, key: deviceKey.privateKey,
                whenAuthenticationFails: .verificationFailed(
                    detail: "The Host did not accept the Device Key; it is not enrolled there yet."
                ))
            onStep(.verify)
            try? await reached.client.close()
            return PairingResult(
                address: reached.address, port: code.port, username: code.username,
                hostKeyFingerprint: reached.presented)
        }

        // Decode-time validation guarantees a 32-byte seed, and CryptoKit
        // accepts any 32 bytes as an Ed25519 seed; an unusable Bootstrap Key
        // is still an authenticate-step failure, not a crash.
        guard
            let bootstrapKey = try? Curve25519.Signing.PrivateKey(
                rawRepresentation: bootstrap.seed)
        else {
            throw PairingCeremonyError.bootstrapRejected
        }

        onStep(.reach)
        let reached = try await reach(
            code: code, key: bootstrapKey, whenAuthenticationFails: .bootstrapRejected)

        onStep(.enroll)
        do {
            try await enroll(deviceKey: deviceKey, over: reached.client)
            try? await reached.client.close()
        } catch {
            try? await reached.client.close()
            throw error
        }

        try Task.checkCancellation()
        onStep(.verify)
        return try await verify(code: code, address: reached.address, deviceKey: deviceKey)
    }

    // MARK: Reach

    private struct ReachedHost {
        let client: SSHClient
        let address: String
        /// Algorithm-aware fingerprint of the key the Host presented;
        /// digest-equal to the pinned one.
        let presented: HostKeyFingerprint
    }

    /// Tries each candidate address in order until one yields an SSH server
    /// presenting the pinned host key and accepting `key`. An address whose
    /// host key does not match the pin is some other machine — skip it and
    /// keep looking. A matching host that rejects the key IS our Host saying
    /// no; that ends the loop with `whenAuthenticationFails`.
    private func reach(
        code: PairingCode,
        key: Curve25519.Signing.PrivateKey,
        whenAuthenticationFails: PairingCeremonyError
    ) async throws -> ReachedHost {
        var attempts: [String] = []
        for address in code.addresses {
            try Task.checkCancellation()
            let validator = PinnedHostKeyValidator(pinned: code.hostKeyFingerprint)
            do {
                let client = try await SSHClient.connect(
                    host: address,
                    port: code.port,
                    authenticationMethod: .ed25519(username: code.username, privateKey: key),
                    hostKeyValidator: .custom(validator),
                    reconnect: .never,
                    connectTimeout: Self.timeAmount(perAddressTimeout))
                guard let presented = validator.presented else {
                    // Unreachable: a completed handshake ran the validator.
                    try? await client.close()
                    attempts.append("\(address): host key was never validated")
                    continue
                }
                return ReachedHost(client: client, address: address, presented: presented)
            } catch {
                // The validator's verdict is the source of truth: NIOSSH may
                // wrap its error on the way out of the handshake.
                if let mismatch = validator.mismatch {
                    attempts.append(
                        "\(address): presented \(mismatch.displayString), not the pinned host key")
                    continue
                }
                switch error {
                case SSHClientError.allAuthenticationOptionsFailed,
                    SSHClientError.unsupportedPasswordAuthentication,
                    SSHClientError.unsupportedPrivateKeyAuthentication,
                    SSHClientError.unsupportedHostBasedAuthentication:
                    throw whenAuthenticationFails
                default:
                    attempts.append("\(address): \(error)")
                }
            }
        }
        throw PairingCeremonyError.hostUnreachable(detail: attempts.joined(separator: "; "))
    }

    // MARK: Enroll

    /// One exec exchange with the Enrollment accept entrypoint: sshd ignores
    /// the requested command (`restrict,command=` forces the accept script)
    /// and the Device Key public line goes in on stdin. The first stdout
    /// line is the whole answer.
    private func enroll(deviceKey: DeviceKey, over client: SSHClient) async throws {
        let submission = deviceKey.authorizedKeysLine(comment: deviceKeyComment) + "\n"
        let responseLine = try await Self.withDeadline(
            enrollTimeout,
            onTimeout: .enrollmentFailed(detail: "the Enrollment exchange timed out")
        ) {
            var stdout = Data()
            var stderr = Data()
            do {
                try await client.withExec("herdr-mobile-enroll") { inbound, outbound in
                    // A fast-failing forced command can close the channel
                    // before this write lands; the read loop still drains
                    // the response or surfaces the failure.
                    try? await outbound.write(ByteBuffer(string: submission))
                    for try await chunk in inbound {
                        switch chunk {
                        case .stdout(let buffer):
                            stdout.append(contentsOf: buffer.readableBytesView)
                            if stdout.contains(0x0A) { return }
                        case .stderr(let buffer):
                            stderr.append(contentsOf: buffer.readableBytesView)
                        }
                    }
                }
            } catch is CancellationError {
                throw CancellationError()
            } catch {
                throw PairingCeremonyError.enrollmentFailed(
                    detail: "\(error); stderr: \(Self.preview(stderr))")
            }
            guard let newline = stdout.firstIndex(of: 0x0A) else {
                throw PairingCeremonyError.enrollmentFailed(
                    detail:
                        "the accept entrypoint closed without answering; stderr: \(Self.preview(stderr))"
                )
            }
            return String(decoding: stdout[stdout.startIndex..<newline], as: UTF8.self)
        }

        switch EnrollmentResponse.parse(line: responseLine) {
        case .enrolled(let fingerprint):
            let expected = HostKeyFingerprint(publicKeyBlob: deviceKey.publicKeyBlob)
            guard fingerprint == expected.displayString else {
                throw PairingCeremonyError.enrollmentFailed(
                    detail: "the Host enrolled a different key: \(fingerprint)")
            }
        case .refused(let refusal):
            throw PairingCeremonyError.enrollmentRefused(refusal)
        case nil:
            throw PairingCeremonyError.enrollmentFailed(
                detail: "unexpected accept response: \(Self.preview(Data(responseLine.utf8)))")
        }
    }

    // MARK: Verify

    /// The reconnect that proves Enrollment took effect: same address, same
    /// pinned host key, Device Key credentials. Only after this succeeds may
    /// the caller persist the Host.
    private func verify(
        code: PairingCode, address: String, deviceKey: DeviceKey
    ) async throws -> PairingResult {
        let validator = PinnedHostKeyValidator(pinned: code.hostKeyFingerprint)
        do {
            let client = try await SSHClient.connect(
                host: address,
                port: code.port,
                authenticationMethod: .ed25519(
                    username: code.username, privateKey: deviceKey.privateKey),
                hostKeyValidator: .custom(validator),
                reconnect: .never,
                connectTimeout: Self.timeAmount(perAddressTimeout))
            guard let presented = validator.presented else {
                try? await client.close()
                throw PairingCeremonyError.verificationFailed(
                    detail: "host key was never validated")
            }
            try? await client.close()
            return PairingResult(
                address: address, port: code.port, username: code.username,
                hostKeyFingerprint: presented)
        } catch let error as PairingCeremonyError {
            throw error
        } catch {
            if let mismatch = validator.mismatch {
                throw PairingCeremonyError.verificationFailed(
                    detail: "the host key changed mid-ceremony to \(mismatch.displayString)")
            }
            switch error {
            case SSHClientError.allAuthenticationOptionsFailed:
                throw PairingCeremonyError.verificationFailed(
                    detail: "the Host did not accept the enrolled Device Key")
            default:
                throw PairingCeremonyError.verificationFailed(detail: "\(error)")
            }
        }
    }

    // MARK: Plumbing

    /// Races `operation` against a deadline. A live exec channel ignores
    /// Swift task cancellation, but cancelling ends the *local* inbound
    /// stream; the exec body then returns and Citadel closes the channel on
    /// the way out (same shape as SSHTransport's request deadline).
    private static func withDeadline<T: Sendable>(
        _ timeout: Duration,
        onTimeout timeoutError: PairingCeremonyError,
        operation: @escaping @Sendable () async throws -> T
    ) async throws -> T {
        try await withThrowingTaskGroup(of: T?.self) { group in
            group.addTask { try await operation() }
            group.addTask {
                try await Task.sleep(for: timeout)
                return nil
            }
            defer { group.cancelAll() }
            guard let finished = try await group.next(), let value = finished else {
                throw timeoutError
            }
            return value
        }
    }

    private static func timeAmount(_ duration: Duration) -> TimeAmount {
        let (seconds, attoseconds) = duration.components
        return .nanoseconds(seconds * 1_000_000_000 + attoseconds / 1_000_000_000)
    }

    private static func preview(_ data: Data) -> String {
        String(decoding: data.prefix(200), as: UTF8.self)
    }
}

/// Validates the Host's key against the fingerprint pinned from the Pairing
/// Code: match proceeds, anything else fails — no TOFU prompt, no store
/// (ADR 0007). The verdict is recorded here because NIOSSH may wrap the
/// promise's failure on its way out of the handshake; callers consult
/// `presented`/`mismatch`, not the thrown error.
final class PinnedHostKeyValidator: NIOSSHClientServerAuthenticationDelegate, Sendable {
    private struct PinMismatch: Error {}

    private let pinned: HostKeyFingerprint
    private let presentedState = Mutex<HostKeyFingerprint?>(nil)
    private let mismatchState = Mutex<HostKeyFingerprint?>(nil)

    init(pinned: HostKeyFingerprint) {
        self.pinned = pinned
    }

    /// The algorithm-aware fingerprint of the accepted host key.
    var presented: HostKeyFingerprint? { presentedState.withLock { $0 } }
    /// The fingerprint of a rejected host key, if validation failed one.
    var mismatch: HostKeyFingerprint? { mismatchState.withLock { $0 } }

    func validateHostKey(
        hostKey: NIOSSHPublicKey, validationCompletePromise: EventLoopPromise<Void>
    ) {
        var buffer = ByteBuffer()
        hostKey.write(to: &buffer)
        let fingerprint = HostKeyFingerprint(publicKeyBlob: Data(buffer.readableBytesView))
        if fingerprint == pinned {
            presentedState.withLock { $0 = fingerprint }
            validationCompletePromise.succeed(())
        } else {
            mismatchState.withLock { $0 = fingerprint }
            validationCompletePromise.fail(PinMismatch())
        }
    }
}

File: android/app/src/main/java/dev/bybee/heeler/pairing/SshPairingConnector.kt
package dev.bybee.heeler.pairing

import dev.bybee.heeler.core.crypto.DeviceKey
import dev.bybee.heeler.core.crypto.HostKeyFingerprint
import dev.bybee.heeler.core.crypto.PairingCode
import dev.bybee.heeler.core.ssh.SshAuthentication
import dev.bybee.heeler.core.ssh.SshConnection
import dev.bybee.heeler.core.ssh.SshException
import dev.bybee.heeler.core.ssh.SshExecChannel
import java.io.ByteArrayOutputStream
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.withTimeout

/**
 * One-shot SSH implementation of Pairing. It intentionally never uses TOFU: each connection is
 * checked against the Pairing Code pin before Bootstrap or Device Key authentication begins.
 */
class SshPairingConnector(
    private val perAddressTimeoutMs: Int = 4_000,
    private val enrollmentTimeoutMs: Long = 20_000,
    private val deviceKeyComment: String = "heeler",
) : PairingConnector {
    init {
        require(perAddressTimeoutMs > 0) { "Per-address timeout must be positive." }
        require(enrollmentTimeoutMs > 0) { "Enrollment timeout must be positive." }
    }

    override suspend fun pair(
        code: PairingCode,
        deviceKey: DeviceKey,
        onStep: (PairingStep) -> Unit,
    ): PairingResult {
        val bootstrap = code.bootstrap
        if (bootstrap == null) {
            onStep(PairingStep.Reach)
            val reached = reach(
                code,
                deviceKey,
                PairingCeremonyError.VerificationFailed(
                    "The Host did not accept this Device Key; it is not enrolled there yet.",
                ),
            )
            closeSilently(reached.connection)
            return PairingResult(
                address = reached.address,
                port = code.port,
                username = code.username,
                hostKeyFingerprint = reached.fingerprint.value,
            )
        }

        val bootstrapSeed = bootstrap.seed
        val bootstrapKey = try {
            DeviceKey.fromSeed(bootstrapSeed) ?: throw PairingCeremonyError.BootstrapRejected
        } finally {
            bootstrapSeed.fill(0)
        }

        onStep(PairingStep.Reach)
        val reached = reach(code, bootstrapKey, PairingCeremonyError.BootstrapRejected)
        try {
            onStep(PairingStep.Enroll)
            enroll(deviceKey, reached.connection)
        } finally {
            closeSilently(reached.connection)
        }

        onStep(PairingStep.Verify)
        return verify(code, reached.address, deviceKey)
    }

    /** A key-match selects an address. Authentication rejection is authoritative, not a failover. */
    private suspend fun reach(
        code: PairingCode,
        identity: DeviceKey,
        authenticationFailure: PairingCeremonyError,
    ): ReachedHost {
        val attempts = ArrayList<String>(code.addresses.size)
        for (address in code.addresses) {
            var presented: HostKeyFingerprint? = null
            val connection = try {
                SshConnection.connect(
                    host = address,
                    port = code.port,
                    timeoutMs = perAddressTimeoutMs,
                ) { hostKey ->
                    HostKeyFingerprint.fromHostKeyBlob(hostKey.rawKey).also { fingerprint ->
                        presented = fingerprint
                    } == code.hostKeyFingerprint
                }
            } catch (error: CancellationException) {
                throw error
            } catch (_: SshException.HostKeyRejected) {
                attempts += "$address: host key did not match the Pairing Code pin"
                continue
            } catch (error: Throwable) {
                attempts += "$address: ${error.detailForUser()}"
                continue
            }

            try {
                authenticate(connection, code.username, identity)
                return ReachedHost(connection, address, checkNotNull(presented))
            } catch (error: CancellationException) {
                closeSilently(connection)
                throw error
            } catch (error: SshException.NativeFailure) {
                closeSilently(connection)
                if (error.operation == "public-key authentication") throw authenticationFailure
                attempts += "$address: ${error.detailForUser()}"
            } catch (error: SshException.Protocol) {
                closeSilently(connection)
                throw authenticationFailure
            } catch (error: SshException) {
                closeSilently(connection)
                attempts += "$address: ${error.detailForUser()}"
            }
        }
        throw PairingCeremonyError.HostUnreachable(attempts.joinToString("; "))
    }

    /** The bootstrap entrypoint accepts precisely one Device Key public-key line. */
    private suspend fun enroll(deviceKey: DeviceKey, connection: SshConnection) {
        val line = deviceKey.openSshPublicKey(deviceKeyComment)
        if (line.contains('\u0000') || line.contains('\n') || line.contains('\r') ||
            line.encodeToByteArray().size >= MAXIMUM_ENROLLMENT_LINE_BYTES
        ) {
            throw PairingCeremonyError.EnrollmentFailed(
                "The Device Key submission was not one bounded line.",
            )
        }
        val response = try {
            readEnrollmentLine(connection, line)
        } catch (error: CancellationException) {
            throw error
        } catch (error: PairingCeremonyError) {
            throw error
        } catch (error: Throwable) {
            throw PairingCeremonyError.EnrollmentFailed(error.detailForUser())
        }

        when (val parsed = EnrollmentResponse.parse(response)) {
            is EnrollmentResponse.Enrolled -> {
                val expected = HostKeyFingerprint.fromHostKeyBlob(deviceKey.publicKeyBlob).value
                if (parsed.deviceKeyFingerprint != expected) {
                    throw PairingCeremonyError.EnrollmentFailed(
                        "The Host enrolled a different Device Key.",
                    )
                }
            }
            is EnrollmentResponse.Refused -> throw PairingCeremonyError.EnrollmentRefused(parsed.refusal)
            null -> throw PairingCeremonyError.EnrollmentFailed("The Enrollment response was not recognized.")
        }
    }

    private suspend fun readEnrollmentLine(connection: SshConnection, publicKeyLine: String): String =
        withTimeout(enrollmentTimeoutMs) {
            val channel = connection.openExec("heeler-enroll")
            try {
                val request = (publicKeyLine + "\n").encodeToByteArray()
                try {
                    channel.write(request)
                } finally {
                    request.fill(0)
                }
                channel.readSingleLine(MAXIMUM_ENROLLMENT_LINE_BYTES)
                    ?: throw PairingCeremonyError.EnrollmentFailed(
                        "The Enrollment entrypoint closed without answering.",
                    )
            } finally {
                closeChannelSilently(channel)
            }
        }

    /** A successful enrollment is only useful after a fresh pinned Device Key authentication. */
    private suspend fun verify(code: PairingCode, address: String, deviceKey: DeviceKey): PairingResult {
        var presented: HostKeyFingerprint? = null
        val connection = try {
            SshConnection.connect(
                host = address,
                port = code.port,
                timeoutMs = perAddressTimeoutMs,
            ) { hostKey ->
                HostKeyFingerprint.fromHostKeyBlob(hostKey.rawKey).also { fingerprint ->
                    presented = fingerprint
                } == code.hostKeyFingerprint
            }
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            throw PairingCeremonyError.VerificationFailed(error.detailForUser())
        }
        try {
            authenticate(connection, code.username, deviceKey)
            return PairingResult(
                address = address,
                port = code.port,
                username = code.username,
                hostKeyFingerprint = checkNotNull(presented).value,
            )
        } catch (error: CancellationException) {
            throw error
        } catch (_: Throwable) {
            throw PairingCeremonyError.VerificationFailed(
                "The Host did not accept the enrolled Device Key.",
            )
        } finally {
            closeSilently(connection)
        }
    }

    private suspend fun authenticate(connection: SshConnection, username: String, key: DeviceKey) {
        val privateKey = key.openSshPrivateKeyPem(deviceKeyComment).encodeToByteArray()
        val publicKey = key.openSshPublicKey(deviceKeyComment).encodeToByteArray()
        try {
            connection.authenticate(
                username,
                SshAuthentication.PublicKey(privateKeyPem = privateKey, publicKeyOpenSsh = publicKey),
            )
        } finally {
            privateKey.fill(0)
            publicKey.fill(0)
        }
    }

    private data class ReachedHost(
        val connection: SshConnection,
        val address: String,
        val fingerprint: HostKeyFingerprint,
    )

    private companion object {
        const val MAXIMUM_ENROLLMENT_LINE_BYTES = 4_096
    }
}

private suspend fun SshExecChannel.readSingleLine(maximumBytes: Int): String? {
    val bytes = ByteArrayOutputStream()
    while (true) {
        val chunk = read(maximumBytes = minOf(1024, maximumBytes)) ?: return null
        for (byte in chunk) {
            if (byte == '\n'.code.toByte()) return bytes.toByteArray().decodeToString()
            if (bytes.size() >= maximumBytes) {
                throw PairingCeremonyError.EnrollmentFailed("The Enrollment response was too large.")
            }
            bytes.write(byte.toInt())
        }
    }
}

private suspend fun closeSilently(connection: SshConnection) {
    try {
        connection.disconnect()
    } catch (_: Throwable) {
        // A completed or failed ceremony retains its original outcome.
    }
}

private suspend fun closeChannelSilently(channel: SshExecChannel) {
    try {
        channel.close()
    } catch (_: Throwable) {
        // The already-read protocol result remains authoritative.
    }
}

private fun Throwable.detailForUser(): String = message?.take(200)?.takeIf(String::isNotBlank)
    ?: javaClass.simpleName
